File: app/app/src/main/java/me/williamhester/reddit/ui/viewholders/TextCommentViewHolder.java
package me.williamhester.reddit.ui.viewholders;

import android.content.res.Resources;
import android.view.View;
import android.widget.TextView;

import me.williamhester.reddit.R;
import me.williamhester.reddit.html.HtmlParseResult;
import me.williamhester.reddit.models.TextComment;
import me.williamhester.reddit.html.HtmlParser;
import me.williamhester.reddit.ui.VotableCallbacks;
import me.williamhester.reddit.ui.text.VotableMovementMethod;

import static butterknife.ButterKnife.findById;

/**
 * Holds a comment that contains text and perform
 */
public class TextCommentViewHolder extends CommentViewHolder<TextComment> {

  private final HtmlParser htmlParser;
  private final TextView author;
  private final TextView flair;
  private final TextView metadata;
  private final TextView body;
  private final View levelIndicator;
  private final VotableMovementMethod linkMovementMethod;

  public TextCommentViewHolder(
      View itemView, HtmlParser htmlParser, final VotableCallbacks clickListener) {
    super(itemView, clickListener);

    this.htmlParser = htmlParser;
    author = findById(itemView, R.id.author);
    flair = findById(itemView, R.id.flair);
    metadata = findById(itemView, R.id.metadata);
    body = findById(itemView, R.id.body);
    levelIndicator = findById(itemView, R.id.level_indicator);
    linkMovementMethod = new VotableMovementMethod(clickListener);
    body.setMovementMethod(linkMovementMethod);
  }

  @Override
  public void setContent(TextComment item) {
    super.setContent(item);
    linkMovementMethod.setVotable(item);

    Resources res = itemView.getResources();

    author.setText(item.getAuthor());
    flair.setText(item.getAuthorFlairText());
    metadata.setText(String.format(res.getString(R.string.comment_metadata),
        item.getScore(),
        calculateShortTime(item.getCreatedUtc())));

    if (item.isHidden()) {
      body.setVisibility(View.GONE);
    } else {
      body.setVisibility(View.VISIBLE);
      HtmlParseResult result = htmlParser.parseHtml(item.getBodyHtml());
      body.setText(result.getText());
    }
  }

  @Override
  View getLevelIndicator() {
    return levelIndicator;
  }
}

File: app/app/src/main/java/me/williamhester/reddit/ui/viewholders/TextCommentViewHolder.kt
package me.williamhester.reddit.ui.viewholders

import android.view.View
import android.widget.TextView

import me.williamhester.reddit.R
import me.williamhester.reddit.models.TextComment
import me.williamhester.reddit.html.HtmlParser
import me.williamhester.reddit.ui.VotableCallbacks
import me.williamhester.reddit.ui.text.VotableMovementMethod

/**
 * Holds a comment that contains text and perform
 */
class TextCommentViewHolder(
    itemView: View,
    private val htmlParser: HtmlParser,
    clickListener: VotableCallbacks
) : CommentViewHolder<TextComment>(itemView, clickListener) {

  private val author: TextView = itemView.findViewById(R.id.author)
  private val flair: TextView = itemView.findViewById(R.id.flair)
  private val metadata: TextView = itemView.findViewById(R.id.metadata)
  private val body: TextView = itemView.findViewById(R.id.body)
  override val levelIndicator: View = itemView.findViewById(R.id.level_indicator)
  private val linkMovementMethod = VotableMovementMethod(clickListener)

  init {
    body.movementMethod = linkMovementMethod
  }

  override fun setContent(item: TextComment) {
    super.setContent(item)
    linkMovementMethod.setVotable(item)

    val res = itemView.resources

    author.text = item.author
    flair.text = item.authorFlairText
    metadata.text = String.format(res.getString(R.string.comment_metadata),
        item.score,
        calculateShortTime(item.createdUtc))

    if (item.isHidden) {
      body.visibility = View.GONE
    } else {
      body.visibility = View.VISIBLE
      val result = htmlParser.parseHtml(item.bodyHtml)
      body.text = result.text
    }
  }
}
